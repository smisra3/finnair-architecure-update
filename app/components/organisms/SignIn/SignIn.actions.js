// @flow
import {
  SIGNIN_FAILURE,
  SIGNIN_REQUEST,
  LOADING_INDICATOR_STATUS,
  VALIDATE_USER_NAME,
  VALIDATE_USER_NAME_SUCCESS,
  SIGNIN_ERRORS,
  UPDATE_LOGIN_FORM
} from './SignIn.constants';
import { signInRequestType } from './types';

export const checkUserValidAction = (userID) => ({
  type: VALIDATE_USER_NAME,
  userID,
});

export const setUserValidAction = (response) => ({
  type: VALIDATE_USER_NAME_SUCCESS,
  response,
});

export const requestSignInAction = (requestData): signInRequestType => ({type: SIGNIN_REQUEST,
  requestData,
});

export const loadingIndicatorAction = (status) => ({
  type: LOADING_INDICATOR_STATUS,
  status
});

export const setErrorsAction = (error) => ({
  type: SIGNIN_ERRORS,
  error
});

export const updateLoginAction = (identifier, value) => {
  console.log(identifier, value);
  return {
    type: UPDATE_LOGIN_FORM,
    identifier,
    value
  }
}
