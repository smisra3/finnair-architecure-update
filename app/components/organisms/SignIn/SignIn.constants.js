export const SIGNIN_REQUEST = 'FIL/SIGNIN/SIGNIN_REQUEST';
export const SIGNIN_ERRORS = 'FIL/SIGNIN/SIGNIN_ERRORS';
export const VALIDATE_USER_NAME = 'FIL/SIGNIN/VALIDATE_USER_NAME';
export const VALIDATE_USER_NAME_SUCCESS = 'FIL/SIGNIN/VALIDATE_USER_NAME_SUCCESS';

export const LOADING_INDICATOR_STATUS = 'FIL/SIGNIN/LOADING_INDICATOR_STATUS';

export const UPDATE_LOGIN_FORM = 'FIL/SIGNIN/UPDATE_LOGIN_FORM';