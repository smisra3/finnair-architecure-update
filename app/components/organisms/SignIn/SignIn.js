// @flow
import { Fragment, Component } from "react";
import styled from "styled-components";
import styles from "./SignIn.style";
import Button from "../../atoms/Button";
import Anchor from "../../atoms/Anchor";
import Input from "../../atoms/Input";
import CheckBox from "../../atoms/CheckBox";
import type { Props, State } from "./types";
import { loginPageLabels } from "../../../locales/en-US";
import { SUPPORT_PAGE } from "../../../routes";
import { checkValidity } from "../../../utils/validation";

const initialState = () => ({
  userNameControl: {
    errorMessage: "",
    rules: [{ required: true }, { minLength: 3 }]
  },
  passwordControl: {
    errorMessage: "",
    rules: [{ required: true }]
  },
  isRememberMeChecked: false,
  isFormSubmitted: false
});

/**
 * Signin Component.
 * This organism will be used for login user accounts.
 * This is state full component, It render a form containing input fields called user
 * Email and Password. There are multiple methods,
 * which handles the user interaction with input fields or validations.
 */
class SignIn extends Component<Props, State> {
  constructor(props) {
    super(props);
    this.loginFormRef = React.createRef();

    /**
     * @type {object}
     * @property {boolean}  isPassword
     * @property {boolean}  isEmail
     * @property {boolean}  error
     * @property {boolean}  fieldLabelEmail
     * @property {boolean}  fieldLabelPassword
     * @property {string}  userName store input field for userName
     * @property {string}  password store input field for password
     */
    this.state = initialState();
  }

  checkUserValidHandler = () => {
    this.validateLoginForm();
    const { checkUserValid, loginForm } = this.props;
    if (loginForm.userName.trim() !== "") {
      checkUserValid(loginForm.userName);
    }
  };

  validateLoginForm = () => {};

  formSubmit = event => {
    event.preventDefault();
    const { submitHandler } = this.props;
    const { userName, password } = this.props.loginForm;
    this.setState({
      isFormSubmitted: true
    });
    submitHandler({
      userName: userName.value,
      password: password.value
    });
  };

  inputChangedHandler = (value, inputIdentifier) => {
    this.props.updateLoginHandler(inputIdentifier, value);
  };

  rememberMeHandler = e => {
    this.setState(prevState => ({
      isRememberMeChecked: !prevState.isRememberMeChecked
    }));
  };

  renderLoginForm = () => {
    const { userName, password } = this.props.loginForm;
    const { isValidUser, isLoaderActive } = this.props;
    const { userNameControl, passwordControl, isRememberMeChecked, isFormSubmitted } = this.state;
    const loginForm = (
      <div className="form-inner-wrapper">
        <Input
          labelContent={loginPageLabels.userNameLabel}
          key={"userName"}
          type="text"
          placeholder=""
          value={userName}
          errorMessage={userNameControl.errorMessage}
          onChange={event => this.inputChangedHandler(event.target.value, "userName")}
          onBlur={event => this.checkUserValidHandler()}
          data-slnm-id="userName"
          isParentSubmitted={isFormSubmitted}
        />
        <p className="login-info">{loginPageLabels.userNameCrnInfoMessage}</p>
        {isValidUser && (
          <Input
            labelContent={loginPageLabels.passwordLabel}
            key={"password"}
            type="password"
            placeholder=""
            value={password}
            errorMessage={passwordControl.errorMessage}
            onChange={event => this.inputChangedHandler(event.target.value, "password")}
            onBlur={event => this.inputChangedHandler(event.target.value, "password")}
            data-slnm-id="password"
          />
        )}
        <div className="bottom-gutter top-gutter">
          <CheckBox
            id="rememberMe"
            isChecked={isRememberMeChecked}
            label={loginPageLabels.rememberMeText}
            onChange={e => this.rememberMeHandler(e)}
          />
        </div>
        <div className="center-align">
          <Button
            loader={isLoaderActive}
            data-slnm-id="loginSubmit"
            className="login-form-submit"
            primary="true"
            fullWidth="true"
            type="submit"
          >
            {!isValidUser ? loginPageLabels.next : loginPageLabels.signInButtonTxt}
          </Button>
        </div>
      </div>
    );
    return loginForm;
  };

  /**
   * render method return markup.
   * Markup contains header h1 tag and form.
   * form has input fields, remember-me checkbox and submit button.
   * @return {ReactElement} markup
   */
  render() {
    const { className, errorMessage, heading } = this.props;
    return (
      <Fragment>
        <form
          className={`${className} sign-in`}
          onSubmit={event => this.formSubmit(event)}
          autoComplete="off"
          ref={this.loginFormRef}
        >
          <div className="help-support circular-buffer center-align">
            <p>
              {loginPageLabels.signInHelpText}{" "}
              <Anchor to={SUPPORT_PAGE}>{loginPageLabels.viewSupportPage}</Anchor>
            </p>
          </div>
          <div className="login-form circular-buffer">
            <h2 className="center-align">WELCOME ON BOARD</h2>
            <h2
              className="sign-in-heading circular-buffer center-align"
              data-slnm-id="signInHeading"
            >
              {heading || loginPageLabels.signInHeading}
            </h2>
            {errorMessage && errorMessage.length ? (
              <div className="response-error sign-in-error">
                <span className="error-message">{errorMessage}</span>
              </div>
            ) : null}
            {this.renderLoginForm()}
          </div>
          <div className="registration-and-info">
            <div className="registration circular-buffer center-align">
              <Anchor secondary="true" to="https://www.finnair.com/gb/gb/corporate/enrollment">
                {loginPageLabels.registerForOnlineAccess}
              </Anchor>
            </div>
            <hr />
            <div className="information circular-buffer">
              <p className="margin-0">
                <strong>{loginPageLabels.importantInformation}</strong>
                {loginPageLabels.importantInformationDetails}
              </p>
            </div>
            <hr />
          </div>
        </form>
      </Fragment>
    );
  }
}

const styledSignIn = styled(SignIn)`
  ${styles};
`;

export default styledSignIn;
