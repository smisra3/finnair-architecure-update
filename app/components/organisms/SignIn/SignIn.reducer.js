// @flow
import {
  SIGNIN_ERRORS,
  LOADING_INDICATOR_STATUS,
  VALIDATE_USER_NAME_SUCCESS,
  UPDATE_LOGIN_FORM
} from './SignIn.constants';
import { updateState } from '../../../utils/utils';

export const initialState = {
  loginForm: {
    userName: '',
    password: ''
  }
};
/**
 * Method to capture error in case of action failure
 * @param state - State
 * @param action - Failure action
 * @return state - error along with current State
 */

const setErrors = (state, { error }) => updateState(state, { errorMessage: error });

const loadingIndicator = (state, {status}) => updateState(state, { isLoaderActive: status });

const setUserValid = (state, { response }) => updateState(state, { isUserValid: response });

const updateLogin = (state, action) => {
  let newState = {...state};
  newState.loginForm[action.identifier] = action.value;
  return newState;
}

const signIn = (state = initialState, action) => {
  switch (action.type) {
    case SIGNIN_ERRORS:
      return setErrors(state, action);
    case LOADING_INDICATOR_STATUS:
      return loadingIndicator(state, action);
    case VALIDATE_USER_NAME_SUCCESS: 
      return setUserValid(state, action)
    case UPDATE_LOGIN_FORM:
      return updateLogin(state, action)
    default:
      return state;
  }
};

export default signIn;
