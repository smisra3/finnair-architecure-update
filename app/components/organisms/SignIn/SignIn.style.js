import { css } from 'styled-components';

const commonStyles = css`
  margin-top: 60px;
  margin-bottom: 60px;
  .circular-buffer{
    padding: 20px;
  }
  .bottom-gutter{
    margin-bottom: 20px;
  }
  .help-support{
    background-color: ${props => props.theme.cautionBackgroundColor};

    p, a{
      font-size: 12px;
    }
  }

  p{
    margin: 0px;
  }

  .login-form{
    background-color: ${props => props.theme.backgroundColorPrimary};
  }
  .response-error {
    font-weight: bold;
    padding-bottom: 15px;
    color:red;
    font-size: 15p;
  }

  .input-header {
    display: block;
    font-size: 12px;
  }

  .login-info{
    margin-bottom: 20px;
    color: ${props => props.theme.textColorSecondaryLight};
  }

  .registration-and-info{
    background-color: ${props => props.theme.backgroundColorSecondary};
  }

  hr{
    border-bottom: 1px solid ${props => props.theme.borderColorSecondary};
    margin:0;
  }

  .information{
    color: ${props => props.theme.textColorSecondaryLight};
    line-height: 18px;
  }

  .footer-links{
    color: ${props => props.theme.textColorSecondaryLight};
    a {
      padding-right: 5px;
      line-height: 16px;
      border-right: 1px solid ${props => props.theme.textColorSecondaryLight};
      margin-right: 5px;
    }
  }

`;

export default commonStyles;

