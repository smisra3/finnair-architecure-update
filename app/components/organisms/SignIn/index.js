export { default } from "./SignIn.container";
export { SignIn } from "./SignIn";
