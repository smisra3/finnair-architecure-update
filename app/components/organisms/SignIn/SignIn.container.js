// @flow
import { Component } from 'react';
import _ from 'lodash';
import { connect } from 'react-redux';
import type { Props, State } from './types';
import { requestSignInAction, setErrorsAction, checkUserValidAction, updateLoginAction } from './SignIn.actions';
import SignIn from './SignIn';

const mapStateToProps = (state): MapStateToProps => ({
  data: _.get(state, ['global']),
  loginForm: _.get(state, ['global', 'signIn', 'loginForm']),
  isValidUser: _.get(state, ['global', 'signIn', 'isUserValid']),
  errorMessage: _.get(state, ['global', 'signIn', 'errorMessage']),
  isLoaderActive: _.get(state, ['global', 'signIn', "isLoaderActive"]),
  deviceType: _.get(state, ['global', 'globalData', 'deviceType']),
});

const mapDispatchToProps = (dispatch): void => ({
  checkUserValid: userID => dispatch(checkUserValidAction(userID)),
  submitHandler: userInputs => dispatch(requestSignInAction(userInputs)),
  resetError: () => dispatch(setErrorsAction({})),
  updateLoginHandler: (identifier, value) => dispatch(updateLoginAction(identifier, value))
});

const SignInContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(SignIn);

export default SignInContainer;