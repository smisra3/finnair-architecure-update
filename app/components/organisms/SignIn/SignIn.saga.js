// @flow
import { call, put, takeLatest } from 'redux-saga/effects';

import { SIGNIN_REQUEST, VALIDATE_USER_NAME, UPDATE_LOGIN_FORM } from './SignIn.constants';
import {
  loadingIndicatorAction,
  setErrorsAction,
  setUserValidAction
} from './SignIn.actions';
import API from '../../../utils/fetch';
import { STATUS_ERROR, STATUS_SUCCESS } from '../../../constants';
import API_URLS from '../../../constants/api/url';
import { setUserStateAction, setAuthenticationAction } from '../../../global/actions';
import { loginResponseError } from '../../../../app/locales/en-US';

export function* validateUserSaga(action): Generator<*, *, *> {
  try {
    yield put(loadingIndicatorAction(true));
    yield put(setUserValidAction(false));
    yield put(setErrorsAction({}));
    const data = yield call(API.fetch, API_URLS.validateUsername);
    if (data.status.toLocaleUpperCase() === STATUS_ERROR) {
      yield put(setErrorsAction(data.message));
      yield put(loadingIndicatorAction(false));
    } else if (data.status.toLocaleUpperCase() === STATUS_SUCCESS) {
      yield put(setUserValidAction(data.data[0].isUserExists));
      yield put(loadingIndicatorAction(false));
    }
  } catch (err) {
    yield put(loadingIndicatorAction(false));
    yield put(setErrorsAction(loginResponseError));
  }
}

export function* userSignInSaga(action): Generator<*, *, *> {
  try {
    yield put(setErrorsAction({}));
    yield put(loadingIndicatorAction(true));
    const data = yield call(API.fetch, API_URLS.login);
    if (data.status.toUpperCase() === STATUS_ERROR) {
      yield put(setAuthenticationAction(false, {}))
      yield put(setErrorsAction(data.message));
      yield put(loadingIndicatorAction(false));
    } else if (data.status.toUpperCase() === STATUS_SUCCESS) {
      yield put(setAuthenticationAction(true, data.data))
      yield put(setUserStateAction(data.data));
      yield put(loadingIndicatorAction(false));
    }
  } catch (err) {
    yield put(loadingIndicatorAction(false));
    yield put(setErrorsAction(loginResponseError));
  }
}

const signInSaga = () => {
  return [takeLatest(VALIDATE_USER_NAME, validateUserSaga),takeLatest(SIGNIN_REQUEST, userSignInSaga)]
}

export default signInSaga;
