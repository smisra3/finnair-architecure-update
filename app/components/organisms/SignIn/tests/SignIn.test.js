import 'jest-styled-components';
import { fromJS } from 'immutable';
import { ThemeProvider } from 'styled-components';
import { takeLatest } from 'redux-saga/effects';
import { shallow } from 'enzyme';
import SignInStyled, { SignInVanilla } from '../SignIn';
import Button from '../../../atoms/Button/Button';
import Input from '../../../atoms/Input';
import signSaga, { userSignIn } from '../SignIn.saga';
import { failureSignIn, requestSignIn } from '../SignIn.actions';
import { SIGNIN_FAILURE, SIGNIN_REQUEST } from '../SignIn.constants';
import Theme from '../../../../styles/theme/';


const stepper = fn => mock => fn.next(mock).value;

const props = {
  submitHandler: jest.fn(),
  signInError: fromJS({}),
};

describe('SignIn with styles', () => {
  it('create snapshot with theme', () => {
    const signIn = shallow(<ThemeProvider theme={Theme}><SignInStyled /></ThemeProvider>);
    expect(signIn).toMatchSnapshot();
  });
});

describe('SignIn Component ', () => {
  let signIn;
  beforeEach(() => {
    signIn = shallow(<SignInVanilla {...props} />);
  });
  test('should perform form submit and set submit state true', () => {
    signIn.setState({
      email: {
        value: 'test@test.com',
        isValid: true,
      },
      password: {
        value: 'test123TEST',
        isValid: true,
      },
    });
    const form = signIn.find('form').at(0);
    form.simulate('submit', { preventDefault: () => {} });
    expect(signIn.state().password.isValid).toEqual(true);
  });

  test('should receieve new props ', () => {
    signIn.setState({
      email: {
        value: 'test@test.com',
        isValid: true,
      },
      password: {
        password: 'test123TEST',
        isValid: true,
      },
      submitted: true,
    });
    signIn.setProps({
      signInError: fromJS({
        failure: [{
          errorCode: 'currentPassword',
          message: 'password doesnt match with email please re-check your password',
        },
        {
          errorCode: 'loginFormError',
          message: 'test login error',
        }],
      }),
    });
    signIn.update();
    expect(signIn.state().password.errorMessage).toEqual('password doesnt match with email please re-check your password');
  });
  test('should render with error true ', () => {
    const form = signIn.find('form').at(0);
    form.simulate('submit', { preventDefault: () => {} });
    expect(signIn.state().email.isError).toEqual(true);
  });
  test('should render with error false and state have email and password  ', () => {
    signIn.setState({
      email: { email: 'test@test.com', value: 'test@test.com' },
      password: { password: 'test123TEST', value: 'test123TEST' },
    });
    const button = signIn.find(Button).at(1);
    button.simulate('click', { preventDefault: () => {} });
    expect(signIn.state().email.value).toEqual('test@test.com');
    expect(signIn.state().password.value).toEqual('test123TEST');
    expect(signIn.state().email.isError).not.toBeDefined();
    expect(signIn.state().password.isError).not.toBeDefined();
  });

  test('should validate email and set error false if email is correct   ', () => {
    const event = { target: { name: 'test', value: 'test@test.com', type: 'email' } };
    const input = signIn.find(Input).at(0);
    input.simulate('change', event);
    input.simulate('blur');
    expect(signIn.state().email.value).toEqual('test@test.com');
    expect(signIn.state().email.isError).toEqual(false);
  });
  test('should validate email and set error true if email is not correct   ', () => {
    const event = { target: { name: 'test', value: 'test', type: 'email' } };
    const input = signIn.find(Input).at(0);
    input.simulate('change', event);
    input.simulate('blur');
    expect(signIn.state().email.value).toEqual('test');
    expect(signIn.state().email.isError).toEqual(true);
  });
  test('should validate password and set error false if password is correct   ', () => {
    const event = { target: { name: 'test', value: 'test12Password', type: 'password' } };
    const input = signIn.find(Input).at(1);
    input.simulate('change', event);
    input.simulate('blur');
    expect(signIn.state().password.value).toEqual('test12Password');
    expect(signIn.state().password.isError).toEqual(false);
  });
  test('should validate password and set error true if password is not correct   ', () => {
    const event = { target: { name: 'test', value: 'testpassword', type: 'password' } };
    const input = signIn.find(Input).at(1);
    input.simulate('change', event);
    input.simulate('blur');
    expect(signIn.state().password.value).toEqual('testpassword');
    expect(signIn.state().password.isError).toEqual(true);
  });
  test('should validate email and set error true if email is blank   ', () => {
    const event = { target: { name: 'test', value: '', type: 'email' } };
    const input = signIn.find(Input).at(0);
    input.simulate('change', event);
    input.simulate('blur');
    expect(signIn.state().email.isError).toEqual(true);
  });
  test('should validate password and set error true if password is blank   ', () => {
    const event = { target: { name: 'test', value: '', type: 'password' } };
    const input = signIn.find(Input).at(1);
    input.simulate('change', event);
    input.simulate('blur');
    expect(signIn.state().password.isError).toEqual(true);
  });
});

describe('SignIn action  ', () => {
  let action;
  test('should return action type failure with data error  ', () => {
    action = failureSignIn('test error');
    expect(action.error).toEqual('test error');
    expect(action.type).toEqual(SIGNIN_FAILURE);
  });
  test('should return action type SIGNIN REQUEST with request data  ', () => {
    action = requestSignIn('test request data');
    expect(action.requestData).toEqual('test request data');
    expect(action.type).toEqual(SIGNIN_REQUEST);
  });
});

describe('SignIn saga testing ', () => {
  let step;
  beforeEach(() => {
    step = stepper(signSaga());
  });

  test('should work SIGNIN_REQUEST correctly', () => {
    expect(step()).toEqual(takeLatest(SIGNIN_REQUEST, userSignIn));
  });
});
