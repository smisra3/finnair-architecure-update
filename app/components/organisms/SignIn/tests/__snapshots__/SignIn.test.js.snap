// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn with styles create snapshot with theme 1`] = `<Connect(Styled(SignIn)) />`;
